docs/proposals/config-management-plugin-v2.md

```diff
@@ -0,0 +1,290 @@
+---
+title: Config-Management-Plugin-Enhancement
+authors:
+  - "@kshamajain99" # Authors' github accounts here.
+sponsors:
+  - TBD        # List all intereste parties here.
+reviewers:
+  - TBD
+approvers:
+  - TBD
+
+creation-date: 2021-03-29
+last-updated: 2021-03-29
+---
+
+# Config Management Plugin Enhancement
+
+We want to enhance config management plugin in order to improve Argo CD operator and end-user experience 
+for using additional tools such as cdk8s, Tanka, jkcfg, QBEC, Dhall, pulumi, etc. 
+
+## Summary
+
+Currently, Argo CD provides first-class support for Helm, Kustomize, Jsonnet/YAML, and Ksonnet. The support includes:
+- Bundled binaries (maintainers periodically upgrade binaries)
+- An ability to override parameters using UI/CLI
+- The applications are discovered in Git repository and auto-suggested during application creation in UI
+- Performance optimizations. Argo CD "knows" when it is safe to generate manifests concurrently and takes advantage of it.
+
+We want to enhance the configuration management plugin so that it can provide similar first-class support for additional 
+tools such as cdk8s, Tanka, jkcfg, QBEC, Dhall, pulumi, etc.
+
+## Motivation
+
+The config management plugin feature should be improved to provide the same level of user experience as 
+for the natively supported tools to the additional tools such as  cdk8s, Tanka, jkcfg, QBEC, Dhall, pulumi, etc., 
+including Argo CD operators as well as end-user experience.
+
+### Goals
+
+The goals for config management plugin enhancement are,
+
+#### Improve Installation Experience
+The current Config Management plugin installation experience requires two changes:
+- An entry in configManagementPlugins in the Argo CD configmap (i.e.  argocd-cm)
+- Either an init container with a volume mount that adds a new binary into Argo CD repo server pod, or a rebuild of the argocd image, which contains the necessary tooling
+
+The problem with this approach is that the process is error-prone, manual, and requires learning from each and every Argo CD administrator. 
+
+The goal is to make additional tools easily accessible for installation to Argo CD operators.
+
+#### Provide Discovery (Auto-selection of Tool)
+For Argo CD’s natively supported config management plugins (Helm, Kustomize, Ksonnet, Jsonnet), Argo CD auto-detects 
+and selects the appropriate tool given only the path in the Git repository. 
+This selection is based on the recognition of well-known files in the directory (e.g. Chart.yaml, kustomization.yaml, etc...). 
+
+Currently, unlike natively supported tools, when a plugin is used, a user needs to explicitly specify the plugin 
+that should be used to render the manifests. As part of the improvements to config management plugins, 
+
+We want to provide the same ability to auto-select the plugin based on recognized files in the path of the git repository.
+
+### Non-Goals
```

Comment:
If I heard it right, we aren't planning on moving the existing/native plugins out of the image as yet.
If yes, we should add that to the non-goal to make it explicit.

Reply:
will do.